src/engine/SpatialJoin.cpp

```diff
@@ -0,0 +1,295 @@
+    auto varColMapLeft = childLeft->getRootOperation()
+          ->getExternallyVisibleVariableColumns();
+    auto varColMapRight = childRight->getRootOperation()
+          ->getExternallyVisibleVariableColumns();
+    ColumnIndex leftJoinCol
+            = varColMapLeft[leftChildVariable.value()].columnIndex_;
+    ColumnIndex rightJoinCol
+            = varColMapRight[rightChildVariable.value()].columnIndex_;
+    for (size_t rowLeft = 0; rowLeft < res_left->size(); rowLeft++) {
+      for (size_t rowRight = 0; rowRight < res_right->size(); rowRight++) {
+        size_t rescol = 0;
+        double dist = computeDist(res_left, res_right, rowLeft, rowRight,
+              leftJoinCol, rightJoinCol);
+        if (dist < maxDist) {
+          result.emplace_back();
+          if (addDistToResult) {
+            result.at(resrow, rescol) = ValueId::makeFromDouble(dist);
+            rescol += 1;
+          }
+          // add columns to result table
+          rescol = addColumns(&result, res_left, resrow, rescol,
+                              rowLeft);
+          rescol = addColumns(&result, res_right, resrow, rescol,
+                              rowRight);
+          resrow += 1;
+        }
+      }
+    }
+    return ResultTable(std::move(result),
+            std::vector<ColumnIndex>{}, LocalVocab{});
+  }
+  AD_THROW("SpatialJoin: this line should never be reached");
+}
+
+
+VariableToColumnMap SpatialJoin::computeVariableToColumnMap() const {
+  VariableToColumnMap variableToColumnMap;
+  auto makeCol = makePossiblyUndefinedColumn;
+  
+  if (!(childLeft || childRight)) {
+    // none of the children has been added
+    variableToColumnMap[leftChildVariable.value()] = makeCol(ColumnIndex{0});
+    variableToColumnMap[rightChildVariable.value()] = makeCol(ColumnIndex{1});
+  } else if (childLeft && !childRight) {
+    // only the left child has been added
+    variableToColumnMap[rightChildVariable.value()] = makeCol(ColumnIndex{1});
+  } else if (!childLeft && childRight) {
+    // only the right child has been added
+    variableToColumnMap[leftChildVariable.value()] = makeCol(ColumnIndex{0});
+  } else {
+    // both children have been added to the Operation
+    auto varColsLeftMap = childLeft->getVariableColumns();
+    auto varColsRightMap = childRight->getVariableColumns();
+    auto varColsLeftVec = copySortedByColumnIndex(varColsLeftMap);
+    auto varColsRightVec = copySortedByColumnIndex(varColsRightMap);
+    size_t index = 0;
+    if (addDistToResult) {
+      variableToColumnMap[Variable{"?distOfTheTwoObjectsAddedInternally"}]
+        = makeCol(ColumnIndex{index});
+      index++;
+    }
+    for (size_t i = 0; i < varColsLeftVec.size(); i++) {
+      variableToColumnMap[varColsLeftVec.at(i).first] = makeCol(ColumnIndex{index});
+      index++;
+    }
+    for (size_t i = 0; i < varColsRightVec.size(); i++) {
+      variableToColumnMap[varColsRightVec.at(i).first] = makeCol(ColumnIndex{index});
+      index++;
+    }
```

Comment:
You do exactly the same code for the left and the right cols, 
factor out the common code into a lambda.